### PR TITLE
Postgres binary

### PR DESCRIPTION
**Pull request by ZuluPro**
@psychok7 If you could test it ?

### CODE REVIEW

**Comment by psychok7** [comment]
@ZuluPro i tested the new connector (dbbackup and dbrestore) and it works.

Don't you prefer to use this connector as the default one for postgres since the other one does not work?

To test this, i had to add this to my settings file:

```
DBBACKUP_CONNECTOR_MAPPING = {
    'django.contrib.gis.db.backends.postgis': 'dbbackup.db.postgresql.PgDumpBinaryConnector',
}
```

File: dbbackup/db/postgresql.py [hunk 1]
```diff
@@ -76,3 +76,48 @@ def _restore_dump(self, dump):
         if self.settings.get('ADMIN_USER'):
             self._enable_postgis()
         return super(PgDumpGisConnector, self)._restore_dump(dump)
+
+
+class PgDumpBinaryConnector(PgDumpConnector):
+    """
+    PostgreSQL connector, it uses pg_dump`` to create an SQL text file
```
Comment (psychok7) [inline dbbackup/db/postgresql.py]: @ZuluPro You probably want to fix this.

**Comment by psychok7** [comment]
@ZuluPro while testing, i found an issue with this. It only seems to happen if i want to restore on a fresh/new database.

```python
root@26a939645aa0:/code# python3 manage.py dbrestore --uncompress --settings=proj.settings.frontend
DJANGO_SETTINGS_MODULE:  proj.settings.frontend
Finding latest backup
Restoring backup for database 'default' and server 'None'
Restoring: default-93dd03e62a8a-2017-05-09-163538.psql.bin.gz
Restore tempfile created: 743.1 KiB
Are you sure you want to continue? [Y/n] y
Traceback (most recent call last):
  File "manage.py", line 22, in <module>
    execute_from_command_line(sys.argv)
  File "/usr/local/lib/python3.5/dist-packages/django/core/management/__init__.py", line 363, in execute_from_command_line
    utility.execute()
  File "/usr/local/lib/python3.5/dist-packages/django/core/management/__init__.py", line 355, in execute
    self.fetch_command(subcommand).run_from_argv(self.argv)
  File "/usr/local/lib/python3.5/dist-packages/django/core/management/base.py", line 283, in run_from_argv
    self.execute(*args, **cmd_options)
  File "/usr/local/lib/python3.5/dist-packages/django/core/management/base.py", line 330, in execute
    output = self.handle(*args, **options)
  File "/usr/local/lib/python3.5/dist-packages/dbbackup/management/commands/dbrestore.py", line 54, in handle
    self._restore_backup()
  File "/usr/local/lib/python3.5/dist-packages/dbbackup/management/commands/dbrestore.py", line 94, in _restore_backup
    self.connector.restore_dump(input_file)
  File "/usr/local/lib/python3.5/dist-packages/dbbackup/db/base.py", line 90, in restore_dump
    result = self._restore_dump(dump)
  File "/usr/local/lib/python3.5/dist-packages/dbbackup/db/postgresql.py", line 122, in _restore_dump
    stdout, stderr = self.run_command(cmd, stdin=dump, env=self.restore_env)
  File "/usr/local/lib/python3.5/dist-packages/dbbackup/db/postgresql.py", line 20, in run_command
    return super(PgDumpConnector, self).run_command(*args, **kwargs)
  File "/usr/local/lib/python3.5/dist-packages/dbbackup/db/base.py", line 144, in run_command
    "Error running: {}\n{}".format(command, stderr.read()))
dbbackup.db.exceptions.CommandConnectorError: Error running:  pg_restore --dbname=postgres --host=proj_postgis --port=5432 --user=postgres --no-password --single-transaction --clean 
b'pg_restore: [archiver (db)] Error while PROCESSING TOC:\npg_restore: [archiver (db)] Error from TOC entry 4549; 2606 72966 FK CONSTRAINT djcelery_taskstate_worker_id_f7f57a05_fk_djcelery_ postgres\npg_restore: [archiver (db)] could not execute query: ERROR:  relation "public.djcelery_taskstate" does not exist\n    Command was: ALTER TABLE ONLY public.djcelery_taskstate DROP CONSTRAINT djcelery_taskstate_worker_id_f7f57a05_fk_djcelery_;\n\n'

```

I read it might be related to the `clean` ( http://stackoverflow.com/a/10120875/977622 )

**Comment by psychok7** [comment]
@ZuluPro any updates on this??

**Comment by ZuluPro** [comment]
@psychok7 Sorry for my latency, I'll take a look this week-end ;)

**Comment by MickaelBergem** [comment]
@ZuluPro checked for basic usage, works fine :)

**Comment by psychok7** [comment]
hey @ZuluPro  any updates on the issue i reported with this PR?? cheers

**Comment by ZuluPro** [comment]
Hello @psychok7 ,
Sorry again I'm overloaded,

I'm looking why your case doesn't work:

I already integrate `--clean` option, it is useless with `pg_dump` binary but used in `pg_restore`. The option is controlled by `DROP` parameter and you seem to use it.

The ALTER TABLE statements seem to be due to creation constraints and pg_dump (or pg_restore) creates them.

Could you try to remove the `DROP` (`clean`) option ? It seems that it is useless with an empty database.

**Comment by psychok7** [comment]
@ZuluPro hi there, i haven't tried removing the `DROP` yet, but even if it works i am not sure it is the right approach because when i backup my database, i do not know "where" i am going to restore it. It could be on a fresh database, or it could be on a database with data in it.

Imagine the scenario where my production machine crashes, if i don't have 2 backups (one with drop and one without) my backup will only work (out of the box) with empty databases (which might not be the case).

Do you have any thoughts on this already?

**Comment by psychok7** [comment]
@ZuluPro i changed the `DROP=True` to  `DROP=False`, and i wrote a custom management command to erase everything in my database before attempting to restore. This way everything seems to work with Postgres binary.

Question, do i still need the `--compress`? or it doesn't make a difference anymore since Postgres already compresses my dumps?

I think this is ready to merge if we add some doc explaining that the user should manually clean the db and set `drop=False`.

**Comment by erback** [comment]
when can we merge this? Really need to be able to use pg_restore

**Comment by ZuluPro** [comment]
@erback 
I'm really sorry for the time this issue and PR take.

Did you try if this PR works for you ?
@psychok7 said it is not, and I clearly not have the time to find his solution.

@psychok7 Don't you think you have trouble with your Django Celery tables ?

**Comment by psychok7** [comment]
@ZuluPro im not sure, but i think you can merge this anyways the workaround for my problem is to erase everything in my database before attempting to restore.

Question, do i still need the --compress? or it doesn't make a difference anymore since Postgres already compresses my dumps?

**Comment by ZuluPro** [comment]
@psychok7
As Postgresql has it's own compression system, `--compress` become useless. Psql does it in an incomparable way for gzip & co.

**Comment by erback** [comment]
@ZuluPro I've tested and it works for me as it seems

**Comment by bartmika** [comment]
+1 for what @thimma11 asked.

**Comment by ZuluPro** [comment]
Hello @thimma11 @bartmika 
Sorry for my latency,

The quick answer:
Yes, it is supposed to be possible.
Yes, you need to specify some conf about the PgBinaryConnector.
The way how it works depends of the Connector, some will include create code in backup, other will specify at restore, etc.

I didn't work on DBbackup since a lot of time,
So please, let me a bit of time give you a good answer and/or add documentation about it.

**Comment by Mee1ad** [comment]
any update?

> Hello @thimma11 @bartmika
> Sorry for my latency,
> 
> The quick answer:
> Yes, it is supposed to be possible.
> Yes, you need to specify some conf about the PgBinaryConnector.
> The way how it works depends of the Connector, some will include create code in backup, other will specify at restore, etc.
> 
> I didn't work on DBbackup since a lot of time,
> So please, let me a bit of time give you a good answer and/or add documentation about it.

any update?

**Pull request merged**